src/vs/workbench/contrib/terminalContrib/accessibility/test/browser/bufferContentTracker.test.ts

```diff
@@ -66,21 +67,56 @@ suite('Buffer Content Tracker', () => {
-		strictEqual(bufferTracker.lines.length, 0);
-		await writeP(xterm.raw, 'abcd\n');
-		await writeP(xterm.raw, 'abcd');
+	test('should not clear the prompt line', async () => {
+		assert.strictEqual(bufferTracker.lines.length, 0);
+		await writeP(xterm.raw, prompt);
 		xterm.clearBuffer();
 		await bufferTracker.update();
-		strictEqual(bufferTracker.lines.length, 1);
+		assert.deepStrictEqual(bufferTracker.lines, [prompt]);
+		assert.strictEqual(bufferTracker.lines.length, 1);
 	});
-	test('should render lines in the viewport', async () => {
-		strictEqual(bufferTracker.lines.length, 0);
-		await writeP(xterm.raw, 'abcd');
+	test('repeated updates should not change the content', async () => {
+		assert.strictEqual(bufferTracker.lines.length, 0);
+		await writeP(xterm.raw, prompt);
+		await bufferTracker.update();
+		assert.deepStrictEqual(bufferTracker.lines, [prompt]);
+		assert.strictEqual(bufferTracker.lines.length, 1);
 		await bufferTracker.update();
-		strictEqual(bufferTracker.lines.length, 1);
+		assert.deepStrictEqual(bufferTracker.lines, [prompt]);
+		assert.strictEqual(bufferTracker.lines.length, 1);
+		await bufferTracker.update();
+		assert.deepStrictEqual(bufferTracker.lines, [prompt]);
+		assert.strictEqual(bufferTracker.lines.length, 1);
+	});
+	test('should add lines in the viewport and scrollback', async () => {
+		await assertBufferExpected(promptPlusData, 38, xterm.raw, bufferTracker);
+	});
+	test('should add lines in the viewport and full scrollback', async () => {
+		await assertBufferExpected(promptPlusData, 1030, xterm.raw, bufferTracker);
+	});
+	test('should always refresh viewport', async () => {
+		await assertBufferExpected(promptPlusData, 6, xterm.raw, bufferTracker);
+		await writeP(xterm.raw, '\x1b[3Ainserteddata');
+		await bufferTracker.update();
+		assert.deepStrictEqual(bufferTracker.lines, [promptPlusData, promptPlusData, `${promptPlusData}inserteddata`, promptPlusData, promptPlusData, promptPlusData].map(s => s.replace(new RegExp(' ', 'g'), '\xA0')));
+	});
+	test('should always refresh viewport with full scrollback', async () => {
+		const content = `${prompt}\r\n`.repeat(1030).trimEnd().replace(new RegExp(' ', 'g'), '\xA0');
+		await assertBufferExpected(prompt, 1030, xterm.raw, bufferTracker);
+		await writeP(xterm.raw, '\x1b[4Ainsertion');
+		await bufferTracker.update();
+		const expected = content.split('\r\n');
+		expected[1025] = `${prompt}insertion`;
+		assert.deepStrictEqual(bufferTracker.lines, expected);
 	});
 });
 
+async function assertBufferExpected(data: string, rows: number, terminal: Terminal, bufferTracker: BufferContentTracker): Promise<void> {
+	const content = `${data}\r\n`.repeat(rows).trimEnd();
+	await writeP(terminal, content);
+	await bufferTracker.update();
+	assert.strictEqual(bufferTracker.lines.length, rows);
+	assert.deepStrictEqual(bufferTracker.lines, content.split('\r\n').map(s => s.replace(new RegExp(' ', 'g'), '\xA0')));
+}
```

Comment:
We should rename this to something like `writeAndAssertBufferState`, or split it into separate write and assert functions. I got a bit confused reading these tests above:

```js
	test('should add lines in the viewport and scrollback', async () => {
		await assertBufferExpected(promptPlusData, 38, xterm.raw, bufferTracker);
	});
	test('should add lines in the viewport and full scrollback', async () => {
		await assertBufferExpected(promptPlusData, 1030, xterm.raw, bufferTracker);
	});
```